Masked behavior

IeuanWalker opened the pull request:
### Description of Change ###

Added MaskedBehavior functionality.
Are tests needed for this? If so not sure how to go about that.

Also, it only works on Entries, not sure if its need for Editors too. If it is needed, what's the best way to do that other than creating a copy of the behaviour and just changing the base control from `Entry` to `Editor`?

### Related issue ###

- Related to issue #179 

### API Changes ###

New behaviour to allow people to mask text for Entry control.
'X' is the user input

```xaml
<Entry Keyboard="Numeric">
    <Entry.Behaviors>
        <behaviors:MaskedBehavior Mask="XX-XX-XX" />
    </Entry.Behaviors>
</Entry>
```

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

### Samples ###
<p align="center">
<img src="https://user-images.githubusercontent.com/6544051/89112982-d134bf80-d462-11ea-8776-d72a013ea2df.gif" height="700" >
</p>

sthewissen commented:
`Editor` and `Entry` both inherit from `Xamarin.Forms.InputView`. Perhaps that might be enough to just target that instead.

AndreiMisiukevich commented:
> `Editor` and `Entry` both inherit from `Xamarin.Forms.InputView`. Perhaps that might be enough to just target that instead.

Just a reminder, that we currently we target XF 3.5 where it can be not working (e.g. for SearchBar)

AndreiMisiukevich commented on XamarinCommunityToolkitSample/Microsoft.Toolkit.Xamarin.Forms.Sample.csproj:
```diff
@@ -42,4 +41,9 @@
+    <Compile Update="Pages\Behaviors\MaskedBehaviorPage.xaml.cs">
+      <DependentUpon>MaskedBehaviorPage.xaml</DependentUpon>
+    </Compile>
+  </ItemGroup>
```
This file changes look suspect for me. I guess something wrong here..

IeuanWalker replied:
Ye I noticed that before creating the PR. But when I tried to change it back I get this error in the sample project - 
![image](https://user-images.githubusercontent.com/6544051/89161296-52c34500-d56a-11ea-8a49-6e485fb1a85d.png)

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,87 @@
+
+                //!important - If user types in masked value, don't masked value
+                if (text.Substring(position.Key, 1) != value)
+                {
```
please don't use parentheses for one line blocks

AndreiMisiukevich replied:
please update it in all files

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace Microsoft.Toolkit.Xamarin.Forms.Behaviors
+{
+    public class MaskedBehavior : Behavior<Entry>
```
Should be inherited from custom "BaseBehavior"

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,87 @@
+        string mask = "";
+        IDictionary<int, char> positions;
+
+        public string Mask
```
I believe it should be a bindable property

AndreiMisiukevich commented on XamarinCommunityToolkit/Microsoft.Toolkit.Xamarin.Forms.csproj:
```diff
@@ -112,4 +111,7 @@
       <Compile Include="**\*.tizen.*.cs" />
     </ItemGroup>
+    <ItemGroup>
+      <Compile Include="Behaviors\MaskedBehavior.cs" />
```
MaskedBehavior.shared.cs

You should create this file in the file system and put it into the proper folder. This csproj shouldn't be modified

IeuanWalker replied:
@AndreiMisiukevich what would be the benefit of making it bindable?

AndreiMisiukevich replied:
Ability to control it from VM via bindings

AndreiMisiukevich replied:
this file must be unchanged

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,87 @@
+            var list = new Dictionary<int, char>();
+            for (var i = 0; i < Mask.Length; i++)
+            {
+                if (Mask[i] != 'X')
```
Mask character should be customizable, so please make a bindable property for it as well

IeuanWalker replied:
ye no problem

IeuanWalker commented:
@sthewissen @AndreiMisiukevich 
I've updated the code so it inherits the BaseBehavior class, And I changed the code to convert the view to InputView rather than the Entry (i.e. `sender as InputView`)

All works fine for entries but when I add an Editor to the page it throws this exception - 
The exception doesn't happen in of the package code, seams like it happen in XF itself, any idea?
```
bindable not an instance of AssociatedType 

 at Xamarin.Forms.Behavior.Xamarin.Forms.IAttachedObject.AttachTo (Xamarin.Forms.BindableObject bindable) [0x0001c] in D:\a\1\s\Xamarin.Forms.Core\Interactivity\Behavior.cs:21 
  at Xamarin.Forms.AttachedCollection`1[T].InsertItem (System.Int32 index, T item) [0x0002b] in D:\a\1\s\Xamarin.Forms.Core\Interactivity\AttachedCollection.cs:61 
  at System.Collections.ObjectModel.Collection`1[T].Add (T item) [0x00020] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/external/corefx/src/Common/src/CoreLib/System/Collections/ObjectModel/Collection.cs:71 
  at Microsoft.Toolkit.Xamarin.Forms.Sample.Pages.Behaviors.MaskedBehaviorPage.InitializeComponent () [0x00012] in D:\WS\GitHub\XamarinToolKit\XamarinCommunityToolkit\XamarinCommunityToolkitSample\obj\Debug\netstandard2.0\Pages\Behaviors\MaskedBehaviorPage.xaml.g.cs:22 
  at Microsoft.Toolkit.Xamarin.Forms.Sample.Pages.Behaviors.MaskedBehaviorPage..ctor () [0x00007] in D:\WS\GitHub\XamarinToolKit\XamarinCommunityToolkit\XamarinCommunityToolkitSample\Pages\Behaviors\MaskedBehaviorPage.xaml.cs:5 
  at (wrapper managed-to-native) System.Reflection.RuntimeConstructorInfo.InternalInvoke(System.Reflection.RuntimeConstructorInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeConstructorInfo.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x00005] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:936 
```

This is what i add to the page -
```xaml
<Editor AutoSize="TextChanges" Keyboard="Text">
            <Editor.Behaviors>
                        <behaviors:MaskedBehavior Mask="(XXX) XXX-XXXX" />
            </Editor.Behaviors>
</Editor>
```

sthewissen commented:
@IeuanWalker Did you also change the type in the AttachedTo / DetachedFrom methods?

IeuanWalker commented:
@sthewissen ye there was no more reference to Entry just ImputView

IeuanWalker commented:
@AndreiMisiukevich @sthewissen 
I think all the code changes request have now been done.
Also works with the editor _(think it just needed a rebuild :/)_

There are 2 bindable properties -

1. Mask, with a default value of string.Empty (e.g. XX-XX-XX)
2. UnMaskedCharacter, with the default value of 'X'

The only issue left is-

- The changes to the csproj. I'm not too sure how to fix this. When I revert the changes to the original I get this error in the Sample project - 
![image](https://user-images.githubusercontent.com/6544051/89212762-f38e2080-d5bb-11ea-910e-3227391e6739.png)

AndreiMisiukevich commented:
> @AndreiMisiukevich @sthewissen
> I think all the code changes request have now been done.
> Also works with the editor _(think it just needed a rebuild :/)_
> 
> There are 2 bindable properties -
> 
> 1. Mask, with a default value of string.Empty (e.g. XX-XX-XX)
> 2. UnMaskedCharacter, with the default value of 'X'
> 
> The only issue left is-
> 
> * The changes to the csproj. I'm not too sure how to fix this. When I revert the changes to the original I get this error in the Sample project -
>   ![image](https://user-images.githubusercontent.com/6544051/89212762-f38e2080-d5bb-11ea-910e-3227391e6739.png)

file name must be MaskedBehaviro.shared.cs

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,90 @@
+            get => (char)GetValue(UnMaskedCharacterProperty);
+            set => SetValue(UnMaskedCharacterProperty, value);
+        }
+
```
Please remove redundant spaces

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,90 @@
+﻿using System;
```
file name must be "MaskedBehavior.shared.cs"

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,90 @@
+            inputView.TextChanged -= OnEntryTextChanged;
+            base.OnDetachingFrom(bindable);
+        }
+        static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue) => ((MaskedBehavior)bindable).SetPositions();
```
please add an empty line above

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.cs:
```diff
@@ -0,0 +1,90 @@
+            base.OnDetachingFrom(bindable);
+        }
+        static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue) => ((MaskedBehavior)bindable).SetPositions();
+        static void OnUnMaskedCharacterPropertyChanged(BindableObject bindable, object oldValue, object newValue) => ((MaskedBehavior)bindable).SetPositions();
```
empty line above here as well

AndreiMisiukevich commented on XamarinCommunityToolkitSample/Pages/Behaviors/MaskedBehaviorPage.xaml.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public partial class MaskedBehaviorPage : BasePage
+    {
+        public MaskedBehaviorPage() => InitializeComponent();
```
```suggestion
        public MaskedBehaviorPage() 
            => InitializeComponent();
```

IeuanWalker commented:
thanks @AndreiMisiukevich, all should be sorted now

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,87 @@
+        IDictionary<int, char> positions;
+
+        public static readonly BindableProperty MaskProperty =
+            BindableProperty.Create(nameof(Mask), typeof(string), typeof(MaskedBehavior), string.Empty, propertyChanged: OnMaskPropertyChanged);
```
Let's leave default value as null if you don't mind :)
```suggestion
            BindableProperty.Create(nameof(Mask), typeof(string), typeof(MaskedBehavior), propertyChanged: OnMaskPropertyChanged);
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,87 @@
+
+        protected override void OnAttachedTo(View bindable)
+        {
+            var inputView = bindable as InputView;
+            inputView.TextChanged += OnEntryTextChanged;
```
```suggestion
            ((InputView)bindable).TextChanged += OnEntryTextChanged;
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,87 @@
+        }
+        protected override void OnDetachingFrom(View bindable)
+        {
+            var inputView = bindable as InputView;
+            inputView.TextChanged -= OnEntryTextChanged;
```
```suggestion
            ((InputView)bindable).TextChanged -= OnEntryTextChanged;
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,87 @@
+        }
+
+        static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue) => ((MaskedBehavior)bindable).SetPositions();
+        static void OnUnMaskedCharacterPropertyChanged(BindableObject bindable, object oldValue, object newValue) => ((MaskedBehavior)bindable).SetPositions();
```
Please add an empty line between static function and move **=>** to the next line

```
void Method()
    => ....
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,87 @@
+
+        void SetPositions()
+        {
+            if (string.IsNullOrEmpty((string)GetValue(MaskProperty)))
```
**(string)GetValue(MaskProperty)**

Should "Mask" property be used?

IeuanWalker commented:
@AndreiMisiukevich Not sure why the pipelines are failing because of the `InputView`, all works fine locally

AndreiMisiukevich commented:
@IeuanWalker that's because we support XF 3.5 for release.

 'InputView' does not contain a definition for 'TextChanged'

jfversluis commented:
Update the dependency, let's just do it. We will need to do it soon enough anway

AndreiMisiukevich commented:
> Update the dependency, let's just do it. We will need to do it soon enough anway

We can work around it, but also we can update dependency

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,88 @@
+            set => SetValue(UnMaskedCharacterProperty, value);
+        }
+
+        protected override void OnAttachedTo(View bindable)
+        {
+            ((InputView)bindable).TextChanged += OnEntryTextChanged;
+            base.OnAttachedTo(bindable);
+        }
+        protected override void OnDetachingFrom(View bindable)
+        {
+            ((InputView)bindable).TextChanged -= OnEntryTextChanged;
+            base.OnDetachingFrom(bindable);
+        }
```
Instead of doing this, just override OnViewPropertyChanged method and check if property name is Entry.TextProperty.PropertyName

IeuanWalker replied:
@AndreiMisiukevich can do, but id also need to do that for the editor and possible the search bar as well. It would also mean anyone who has built their own control that inherits input view wouldn't able to use it (thinking packages like Synctusion, Grail etc)

AndreiMisiukevich replied:
That's just a string check :) Entry.TextProperty.PropertyName == "Text") You can do it as InputView.TextProperty.PropertyName

IeuanWalker replied:
@AndreiMisiukevich oh ok, would i need to change something in the `OnEntryTextChanged` method? - 
```csharp
var inputView = sender as InputView;
var text = inputView.Text;
```

AndreiMisiukevich replied:
var text = ((InputView)View).Text;

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            base.OnViewPropertyChanged(sender, e);
+
+            if(e.PropertyName.Equals("InputView.TextProperty.PropertyName")) return;
```
```suggestion
            if(e.PropertyName == InputView.TextProperty.PropertyName) 
                return;
```

IeuanWalker replied:
@AndreiMisiukevich Done. Is there a difference between `.Equals` and `==` or is it just a style choice?

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            base.OnViewPropertyChanged(sender, e);
+
+            if(e.PropertyName != "Text") return;
```
```suggestion
            if (e.PropertyName != nameof(InputView.Text)) 
                return;
```

AndreiMisiukevich commented on XamarinCommunityToolkit/Behaviors/MaskedBehavior.shared.cs:
```diff
@@ -0,0 +1,82 @@
+            var inputView = (InputView)View;
+            var text = inputView.Text;
+
+            if (string.IsNullOrWhiteSpace(text) || positions == null) return;
```
```suggestion
            if (string.IsNullOrWhiteSpace(text) || positions == null) 
                return;
```

AndreiMisiukevich commented:
I did myself

AndreiMisiukevich commented:
> Build failing

That's because of XF 3.5 Error CS0117: 'InputView' does not contain a definition for 'Text'

AndreiMisiukevich commented:
First of all, we should decide if we upgrade XF version or not

jfversluis commented on XamarinCommunityToolkitSample/Pages/Behaviors/MaskedBehaviorPage.xaml:
```diff
@@ -0,0 +1,88 @@
+            </Entry>
+            <Label Padding="10,0,10,50"
+                   FontSize="Small"
+                   Text="e.g. Sort code 'XX-XX-XX'" />
```
Would you be able to put the texts in the resource file that's there so it will work for the multi-language things?

IeuanWalker replied:
@jfversluis  Ye sure is there any naming conversion for the resources?

jfversluis replied:
@Char0394 ?

IeuanWalker commented:
Thanks @felipebaltazar, they are situations I didn't think of testing. I've added you as a collaborator to my fork.
Feel free to make the changes to fix this.

If not ill prob take a look in a few days

felipebaltazar commented:
@IeuanWalker  Ive updated with my suggestion, feel free to ask or disagree any of this changes
https://github.com/xamarin/XamarinCommunityToolkit/pull/180/commits/0a2e8bbbc512d56c921665744005b9c4eff759e9

AndreiMisiukevich commented:
> Awesome work on this!
> 
> Would you please rename the namespace to the new one? Thanks :)
> 
> Also, please let me know when you feel this is ready for reviewing.

You forgot to mention that the formatting should be updated as well since we moved from spaces to tabs

IeuanWalker commented:
Thanks @felipebaltazar all looks good to me.

@jfversluis @AndreiMisiukevich Namespaces has been updated and all text has been moved to the resources

AndreiMisiukevich commented:
please also change spaces to tabs

Edit->Format->Format document (all new files)

jfversluis commented on XamarinCommunityToolkitSample.iOS/Microsoft.Toolkit.Xamarin.Forms.Sample.iOS.csproj:
```diff
@@ -0,0 +1,169 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
What's happening with this file? Looks like it has the Microsoft.Toolkit... name. So probably some wrong merge? :)

IeuanWalker replied:
@jfversluis strange, I've now deleted that file, also think I've fixed the the formatting

jfversluis commented:
Sorry trying to (mis)use your PR to try and incorporate a dotnet format thing :)

Your changes seem all good though! I'll merge soon

IeuanWalker commented:
> Sorry trying to (mis)use your PR to try and incorporate a dotnet format thing :)
> 
> Your changes seem all good though! I'll merge soon

Lol no problem, thanks.

This may useful (haven't tried it myself yet) - https://github.com/nektos/act

AndreiMisiukevich commented:
@jfversluis I've just noticed in this PR spaces were used instead of tabs
now we have an inconsistent indentation

jfversluis commented:
I know. I'm working on it
